File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxL = 0
        charset = set()
        l = 0
        for r in range(len(s)):
            while s[r] in charset:
                charset.remove(s[l])
                l+=1
            charset.add(s[r])
            maxL = max(maxL, r-l+1)

        return maxL
        

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int ans = 0;
        int l = 0, r=0;
        Set<Character> set = new HashSet<>();
        while(r < s.length()){
            if(set.contains(s.charAt(r))){
                while(s.charAt(l)!=s.charAt(r)){
                    set.remove(s.charAt(l));
                    l++;
                }
                l++;
            }else{
                set.add(s.charAt(r));
                ans = Math.max(ans, (r-l+1));
            }
            r++;
        }
        return ans;
    }
}